src/_purryOrderRules.ts

```diff
@@ -0,0 +1,147 @@
+import type { NonEmptyArray } from './_types';
+
+export type OrderRule<T> = OrderProjection<T> | OrderPair<T>;
+export type CompareFunction<T> = (a: T, b: T) => number;
+
+type OrderPair<T> = readonly [
+  projector: OrderProjection<T>,
+  direction: Direction,
+];
+
+type OrderProjection<T> = (x: T) => Comparable;
+
+type Comparable = ComparablePrimitive | { valueOf(): ComparablePrimitive };
+type ComparablePrimitive = number | string | boolean;
+
+const ALL_DIRECTIONS = ['asc', 'desc'] as const;
+type Direction = (typeof ALL_DIRECTIONS)[number];
+
+const COMPARATOR = {
+  asc: <T>(x: T, y: T) => x > y,
```

Comment:
**nit:** `T extends Comparable`?

Reply:
If typescript is happy, I'm happy. It's best not to limit the implementation if it doesn't need to be limited; that way, a future change that would want to use these comparators with something other than Comparable would be able to do so without needing to change this code.